src/index.js

```diff
@@ -1,27 +1,188 @@
+          needsWrapper
+        } = this.data;
+
+        const eager = needsWrapper || containerNeedsWrapper;
+
+        // Filter out empty children, and transform JSX expressions
+        // into normal expressions.
+        const openingElement = node.openingElement;
+        const closingElement = node.closingElement;
+        const children = buildChildren(t, scope, file, node.children, eager);
+
+        let elements = [ openingElement, ...children ];
+        if (closingElement) { elements.push(closingElement); }
+
+        // If we're inside a JSX node, flattening expressions
+        // may force us into an unwanted function scope.
+        if (t.isJSXElement(parent)) {
+          // If the parent needs to be wrapped, we need to place our eager
+          // child expressions outside the wrapper.
+          if (containerNeedsWrapper) {
+            elements = filterEagerDeclarators(t, elements, eagerDeclarators);
+          }
+          return elements;
+        }
+
+        // Expressions Containers must contain an expression and not statements.
+        // This will be flattened out into statements later.
+        if (containingJSXElement && !needsWrapper) {
+          const sequence = t.sequenceExpression(elements);
+          // Mark this sequence as a JSX Element so we can avoid an unnecessary
+          // renderArbitrary call.
+          sequence._wasJSX = true;
+          return sequence;
+        }
+
+        // Transform (recursively) any sequence expressions into a series of
+        // statements.
+        elements = flattenExpressions(t, elements);
+
+        // If we need to wrap this JSX element in a wrapper, we must place
+        // any eagerly evaluated child expressions outside the wrapper.
+        if (needsWrapper) {
+          elements = filterEagerDeclarators(t, elements, eagerDeclarators);
+        }
+
+
+        if (eagerDeclarators.length && !containingJSXElement) {
+          // Find the closest statement, and insert our eager declarations
+          // before it.
+          const parentStatement = this.findParent((path) => {
+            return path.isStatement();
+          });
+          const declaration = t.variableDeclaration("let", eagerDeclarators);
+          const [path] = parentStatement.insertBefore(declaration);
+          // Add our eager declarations to the scopes tracked bindings.
+          scope.registerBinding("let", path);
+        }
+
+        // Ensure the last statement returns the DOM element.
+        elements = statementsWithReturnLast(t, elements);
+
+        if (needsWrapper) {
+          // Create a wrapper around our element, and mark it as a one so later
+          // child expressions can identify and "render" it.
+          const jsxWrapperRef = injectJSXWrapper(t, file);
+          const wrapper = toFunctionCall(t, jsxWrapperRef, [
+            t.functionExpression(null, [], t.blockStatement(elements))
+          ]);
+          wrapper._wasJSX = true;
```

Comment:
Is it necessary? I was trying to use `path.setData`, but we won't have access to the child's path in the containing JSX Element.